File: summary/rust/13_bfs.rs
use std::collections::{VecDeque, HashMap};

fn breadth_first_search(graph: &HashMap<char, Vec<char>>, start_vertex: char) -> Vec<char> {
    let mut queue = VecDeque::new();
    let mut visited = HashMap::new();
    let mut order_of_visit = Vec::new();

    queue.push_back(start_vertex);
    visited.insert(start_vertex, true);

    while let Some(current_vertex) = queue.pop_front() {
        order_of_visit.push(current_vertex);

        if let Some(neighbors) = graph.get(&current_vertex) {
            for &neighbor in neighbors {
                if !visited.contains_key(&neighbor) {
                    queue.push_back(neighbor);
                    visited.insert(neighbor, true);
                }
            }
        }
    }

    order_of_visit
}

fn main() {
    let mut graph = HashMap::new();
    graph.insert('A', vec!['B', 'C']);
    graph.insert('B', vec!['A', 'D', 'E']);
    graph.insert('C', vec!['A', 'F']);
    graph.insert('D', vec!['B']);
    graph.insert('E', vec!['B', 'F']);
    graph.insert('F', vec!['C', 'E']);

    let visit_order = breadth_first_search(&graph, 'A');
    println!("Visit order: {:?}", visit_order);
}

File: summary/elixir/13_bfs.exs
defmodule BreadthFirstSearch do
  def bfs(graph, start_vertex) do
    queue = :queue.from_list([start_vertex])
    visited = MapSet.new([start_vertex])
    do_bfs(graph, queue, visited, [])
  end

  defp do_bfs(graph, queue, visited, order) do
    case :queue.out(queue) do
      {{:value, vertex}, rest} ->
        {new_queue, new_visited, new_order} =
          graph[vertex]
          |> Enum.reduce({rest, visited, [vertex | order]}, fn neighbor, {q, v, o} ->
            if MapSet.member?(v, neighbor) do
              {q, v, o}
            else
              {:queue.in(neighbor, q), MapSet.put(v, neighbor), o}
            end
          end)

        do_bfs(graph, new_queue, new_visited, new_order)

      {:empty, _} ->
        Enum.reverse(order)
    end
  end
end

# 使用例
graph = %{
  "A" => ["B", "C"],
  "B" => ["A", "D", "E"],
  "C" => ["A", "F"],
  "D" => ["B"],
  "E" => ["B", "F"],
  "F" => ["C", "E"]
}

IO.inspect(BreadthFirstSearch.bfs(graph, "A"))
